instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+	return func(ctx context.Context, event interface{}, takesContext bool) []reflect.Value {
+
+		ctx, span := tracingBegin(ctx)
+		defer tracingEnd(ctx, span)
+
+		handler := reflect.ValueOf(handlerFunc)
+		var args []reflect.Value
+		if takesContext {
+			args = append(args, reflect.ValueOf(ctx))
+		}
+		if eventExists(event) {
+			args = append(args, reflect.ValueOf(event))
+		}
+
+		response := handler.Call(args)
+
+		return response
+	}
+}
+
+// Determine if an interface{} is nil or the
+// if the reflect.Value of the event is nil
+func eventExists(event interface{}) bool {
+	if event == nil {
+		return false
+	}
+
+	// reflect.Value.isNil() can only be called on
+	// Values of certain Kinds. Unsupported Kinds
+	// will panic rather than return false
+	switch reflect.TypeOf(event).Kind() {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
+		return !reflect.ValueOf(event).IsNil()
+	}
+	return true
+}
+
+type wrappedHandler struct {
+	handler lambda.Handler
+}
+
+// Compile time check our Handler implements lambda.Handler
+var _ lambda.Handler = wrappedHandler{}
+
+// Invoke adds OTel span surrounding customer Handler invocation
+func (h wrappedHandler) Invoke(ctx context.Context, payload []byte) ([]byte, error) {
+
+	ctx, span := tracingBegin(ctx)
+	defer tracingEnd(ctx, span)
+
+	response, err := h.handler.Invoke(ctx, payload)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+// HandlerWrapper Provides a Handler which wraps customer Handler with OTel Tracing
+func HandlerWrapper(handler lambda.Handler) lambda.Handler {
+	return wrappedHandler{handler: handler}
+}
+
+// Logic to start OTel Tracing
+func tracingBegin(ctx context.Context) (context.Context, trace.Span) {
+	// Add trace id to context
+	xrayTraceId := os.Getenv("_X_AMZN_TRACE_ID")
+	mc := mapCarrier{}
+	mc.Set("X-Amzn-Trace-Id", xrayTraceId)
+	propagator := xray.Propagator{}
+	ctx = propagator.Extract(ctx, mc)
+
+	// If tracer provider initialization failed we
+	// will attempt to initialize once per invocation
+	if tp == nil {
+		initTracerProvider()
+	}
+
+	// if tracer provider successfully initializes then
+	// we add tracing, otherwise do customer business
+	// logic with no tracing
+	if tp != nil {
+		// Get a named tracer with package path as its name.
+		tracer := tp.Tracer(tracerName)
+
+		var span trace.Span
+		spanName := os.Getenv("AWS_LAMBDA_FUNCTION_NAME")
+
+		var attributes []attribute.KeyValue
+		lc, ok := lambdacontext.FromContext(ctx)
+		if !ok {
+			errorLogger.Println("failed to load lambda context from context, ensure tracing enabled in Lambda")
+		}
+		if lc != nil {
+			ctxRequestID := lc.AwsRequestID
+			attributes = append(attributes, attribute.KeyValue{Key: semconv.FaaSExecutionKey, Value: attribute.StringValue(ctxRequestID)})
+
+			// Resource attrs added as span attr due to static tp
+			// being created without meaningful context
```

Comment:
Is there no mechanism for retrieving these values prior to the first invocation so that they can be added to the `Resource` where they belong?  If not, can they at least be persisted somewhere to avoid reconstructing them on every request?

Reply:
Unfortunately these resources are not available until an invocation. I have now stored them to avoid computing them every time.